### PR TITLE
[Feat] Try to compatible with Keras 3 optimizer design and support CUDNN 9.0+.

### PR DESCRIPTION
**Pull request by MoFHeka**
# Description
Compatibility updates and some compiling log optimizations have been made.
[feat] Competible with Keras3 optimizer style.
[fix] Suppress nodiscard warnings and unused warnings by adding a new LOG_IF_ERROR macro and using proper api function.
[feat] Update the config file for new CUDNN release. Now support CUDNN 9.0+
[fix] self.params may not with saveable attribution.

## Type of change

- [x] Bug fix
- [ ] New Tutorial
- [ ] Updated or additional documentation
- [ ] Additional Testing
- [x] New Feature

# Checklist:

- [x] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/recommenders-addons/blob/master/CONTRIBUTING.md#coding-style)
    - [x] By running yapf
    - [x] By running clang-format
- [ ] This PR addresses an already submitted issue for TensorFlow Recommenders-Addons
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

# How Has This Been Tested?

use default keras.optimizers in Keras 2.12+ instead of keras.legacy.optimizers.

### CODE REVIEW

File: tensorflow_recommenders_addons/dynamic_embedding/core/kernels/dynamic_partition_op_gpu.cu.cc [hunk 1]
```diff
@@ -120,9 +120,9 @@ void MoveValues(const GPUDevice& d, int32* keys, int32* values, int32* num_runs,
 }
 
 struct IdentityOp {
-  __device__ int32 __forceinline__ operator()(const int32& a) const {
-    return a;
-  }
+  // __device__ int32 __forceinline__ operator()(const int32& a) const {
```
Comment (poinwater) [inline tensorflow_recommenders_addons/dynamic_embedding/core/kernels/dynamic_partition_op_gpu.cu.cc:123]: Why are these lines commented out? Could you please add a note explaining the reason for commenting them, to aid in future maintenance?

File: tensorflow_recommenders_addons/dynamic_embedding/core/utils/utils.h [hunk 2]
```diff
@@ -42,4 +42,24 @@ This code is for compatibility.*/
 }  // namespace recommenders_addons
 }  // namespace tensorflow
 
+// For propagating errors when calling a function but not return status.
+#if TF_VERSION_INTEGER >= 2130
```
Comment (poinwater) [inline tensorflow_recommenders_addons/dynamic_embedding/core/utils/utils.h:46]: The code is duplicated here, will this work?
```
#define LOG_IF_ERROR(...)                  \
  do {                                     \
    const auto _status = (__VA_ARGS__);    \
    if (TF_PREDICT_FALSE(!_status.ok())) { \
      #if TF_VERSION_INTEGER >= 2130       \
        MAYBE_ADD_SOURCE_LOCATION(_status) \
        LOG(ERROR) << _status.message();   \
      #else                                \
        LOG(ERROR) << _status.error_message(); \
      #endif                               \
    }                                      \
  } while (0)
```

File: tensorflow_recommenders_addons/dynamic_embedding/core/kernels/dynamic_partition_op_gpu.cu.cc (hunk 1 above)
Comment (MoFHeka) [reply tensorflow_recommenders_addons/dynamic_embedding/core/kernels/dynamic_partition_op_gpu.cu.cc:123]: It looks like I'd be better off removing them. I didn't make it because the commented code is copied from the tensorflow source code.

File: tensorflow_recommenders_addons/dynamic_embedding/core/kernels/lookup_impl/lookup_table_op_hkv.h [hunk 3]
```diff
@@ -552,8 +553,8 @@ class TableWrapper {
     } else {
       wfile.reset(new RandomKVFile<K, V, uint64_t>(
           fs, filepath, dim, buffer_size, append_to_file));
-      status = reinterpret_cast<RandomKVFile<K, V, uint64_t>*>(wfile.get())
-                   ->open(keyfile, valuefile, "wb");
+      status.Update(reinterpret_cast<RandomKVFile<K, V, uint64_t>*>(wfile.get())
```
[the remark this answers is not in the document]
Comment (MoFHeka) [reply tensorflow_recommenders_addons/dynamic_embedding/core/kernels/lookup_impl/lookup_table_op_hkv.h:556]: absl "status" feature override the "=". It wouldn't in-place update the origin status object but create a new one and assign. That would cause a compiler warning--the status was not used.

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py [hunk 4]
```diff
@@ -329,7 +329,7 @@ def _update_step_fn(var, grad):
           with ops.control_dependencies(_before):
             _apply_op = _update_step_fn(var, grad)
 
-          with ops.control_dependencies([_apply_op]):
+          with ops.control_dependencies([var]):
```
[the remark this answers is not in the document]
Comment (MoFHeka) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py:332]: At new Keras 3 optimizer, there's not returning object from OPT update_step function. It just a multiple operators combination for XLA jit compiling. So _apply_op will be a None when use Keras>=2.13 default optimizer.

File: tensorflow_recommenders_addons/dynamic_embedding/core/utils/utils.h (hunk 2 above)
Comment (MoFHeka) [reply tensorflow_recommenders_addons/dynamic_embedding/core/utils/utils.h:46]: if I am not mistaken, the two compatibility define macro for different TF version will follow the the criterion "if TF_VERSION_INTEGER >= 2130" be compiled.

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (MoFHeka) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py:332]: This is related to the update logic of TFRA. var variable(TrainableWrapper) is a training agent for TF. 
Because TFRA does not have its own optimizer but reuses TF's native optimizer, TF cannot directly update the values in the KV table like other open source projects. Instead, TF will update a temporary variable(TrainableWrapper). After the gradient is updated to var, the new values of var will be inserted back into KV table.

**Comment by MoFHeka** [comment]
Yes, of course. 在 2024年3月22日，05:17，j ***@***.***> 写道：﻿
@jq commented on this pull request.



In tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py:
> @@ -329,7 +329,7 @@ def _update_step_fn(var, grad):
           with ops.control_dependencies(_before):
             _apply_op = _update_step_fn(var, grad)
 
-          with ops.control_dependencies([_apply_op]):
+          with ops.control_dependencies([var]):

👍🏻 thanks for the info.  one last question,  remove _apply_op = from below seems valid?
with ops.control_dependencies(_before):
_apply_op = _update_step_fn(var, grad)

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (MoFHeka) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py:332]: > 👍🏻 thanks for the info. one last question, remove `_apply_op =` from below seems valid? with ops.control_dependencies(_before): _apply_op = _update_step_fn(var, grad)

Yes, of course.

File: tensorflow_recommenders_addons/dynamic_embedding/core/utils/utils.h [hunk 5]
```diff
@@ -42,4 +42,24 @@ This code is for compatibility.*/
...
+#define LOG_IF_ERROR(...)                  \
```
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/core/utils/utils.h:47]: Better name it `TFRA_LOG_IF_ERROR` because the macro has no name scope.

Comment (MoFHeka) [reply tensorflow_recommenders_addons/dynamic_embedding/core/utils/utils.h]: fixed

**Review by rhdong — APPROVED** [review]
LGTM
